KactusAPI/expressions/SystemVerilogExpressionParser.cpp

```diff
@@ -353,19 +405,41 @@ bool SystemVerilogExpressionParser::isUnaryOperator(QString const& token) const
 // Function: SystemVerilogExpressionParser::isBinaryOperator()
 //-----------------------------------------------------------------------------
 bool SystemVerilogExpressionParser::isBinaryOperator(QString const& token) const
-{    
+{
     static const QVector<QString> binaryOperators({ QStringLiteral("+"),
-        QStringLiteral("-"), QStringLiteral("*"), QStringLiteral("/"), 
+        QStringLiteral("-"), QStringLiteral("*"), QStringLiteral("/"),
         QStringLiteral("%"), QStringLiteral("**"), QStringLiteral("<<"),
         QStringLiteral(">>"), QStringLiteral("<"), QStringLiteral("<="),
-        QStringLiteral(">"), QStringLiteral(">="), QStringLiteral("!="), 
-        QStringLiteral("!=="), QStringLiteral("=="), QStringLiteral("==="), 
-        QStringLiteral("&"), QStringLiteral("&&"), QStringLiteral("|"), 
+        QStringLiteral(">"), QStringLiteral(">="), QStringLiteral("!="),
+        QStringLiteral("!=="), QStringLiteral("=="), QStringLiteral("==="),
+        QStringLiteral("&"), QStringLiteral("&&"), QStringLiteral("|"),
         QStringLiteral("||"), QStringLiteral("^"), QStringLiteral("$pow")});
 
     return binaryOperators.contains(token);
 }
 
+//-----------------------------------------------------------------------------
+// Function: SystemVerilogExpressionParser::solveTrinary()
+//-----------------------------------------------------------------------------
+QString SystemVerilogExpressionParser::solveTrinary(QString const& operation, QString const& leftTerm,
+    QString const& middleTerm, QString const& rightTerm) const
+{
+    if (isLiteral(leftTerm) && operation.compare(QLatin1String(":")) == 0 )
+    {
+        qreal leftOperand = leftTerm.toDouble();
+        if (leftOperand > 0)

```

Comment:
Here leftOperand (the condition) should be compared if it is equal to 1.

Reply:
> Here leftOperand (the condition) should be compared if it is equal to 1 (true).

If I change it to compare with 1, there will be 3 test failures.

E.g.: ``(3+2)?((3+3)?(4+7):1):0``
``3+2`` will got ``5``, so here needs to be greater than 0 to be true.


```c
        qreal leftOperand = leftTerm.toDouble();
        if (leftOperand == 1)
        {
            return middleTerm;
        }
        else
        {
            return rightTerm;
        }
```

```bash
FAIL!  : tst_SystemVerilogExpressionParser::testTrinaryOperations(Trinary (3+2)?((3+3)?(4+7):1):0 should be 11) Compared values are not the same
   Actual   (parserResult)  : "0"
   Expected (expectedResult): "11"
   Loc: [tst_SystemVerilogExpressionParser.cpp(102)]
PASS   : tst_SystemVerilogExpressionParser::testTrinaryOperations(Trinary (1>2)?((3+3)?(4+7):1):0 should be 0)
FAIL!  : tst_SystemVerilogExpressionParser::testTrinaryOperations(Trinary 3+2?3+3?4+7:1:0 should be 11) Compared values are not the same
   Actual   (parserResult)  : "0"
   Expected (expectedResult): "11"
   Loc: [tst_SystemVerilogExpressionParser.cpp(102)]
PASS   : tst_SystemVerilogExpressionParser::testTrinaryOperations(Trinary 1>2?3+3?4+7:1:0 should be 0)
PASS   : tst_SystemVerilogExpressionParser::testTrinaryOperations(Trinary 1?11:2?12:0 should be 11)
PASS   : tst_SystemVerilogExpressionParser::testTrinaryOperations(Trinary 1?11:(2?12:0) should be 11)
FAIL!  : tst_SystemVerilogExpressionParser::testTrinaryOperations(Trinary 1?12?1:0:22 should be 1) Compared values are not the same
   Actual   (parserResult)  : "0"
   Expected (expectedResult): "1"
   Loc: [tst_SystemVerilogExpressionParser.cpp(102)]
```

Reply:
Yes, then 3 tests will fail, which indicates these cases are testing incorrect behavior. Even if the condition is a mathematical calculation, the end result is a boolean condition (True/False). Only 1 is interpret as true, all other values are interpret as false.

Reply:
![Screenshot from 2022-09-25 01-08-45](https://user-images.githubusercontent.com/394260/192110340-9343a666-31d0-4e4e-8c59-8b278f908f1f.png)

I may not understand correctly, I need a little help here. I tested the three expressions on the picture, and if I want to make leftOperand reasonable, it seems I should use ``.toBool()``?

```c
        bool leftOperand = leftTerm.toBool();
        if (leftOperand == true)
        {
            return middleTerm;
        }
        else
        {
            return rightTerm;
        }

```

Reply:
I tested this trinary operators in ``synopsys vcs``/``cadence irun``/``iverilog``/``verilator``, and found that the condition of ``<condition>?`` can be boolean arithmetic, integer arithmetic and floating-point arithmetic, and it is true if it is not equal to 0 (negative number is also true). 

And ``bool leftOperand = leftTerm.toBool();`` also behaves this way. Should I go for this way?

Reply:
Yes, comparison of floating points can be a bit tricky. That could be reason why the other tools compare against 0, not 1.
`if (condition.toDouble() != 0)` seems to work properly, so shall we go with that?